cmd/boulder-janitor/janitor/job.go

```diff
@@ -211,19 +221,27 @@ LIMIT :limit`, j.expiresColumn, j.table)
 	}
 	lastID := startID
 	rows := 0
+	oooRows := 0
 	cutoff := j.clk.Now().Add(-j.purgeBefore)
 	for _, v := range data {
 		// We check for the expiration in code rather than doing so in the
 		// database query as it allows us to impose a bound on the number
-		// of rows that the database will examine. If a row is return that
-		// has an expiry after the cutoff all of the successive rows
-		// should also have an expiry after the cutoff so we break from
-		// the loop and ignore the rest of the results.
+		// of rows that the database will examine. If a row is returned that
+		// has an expiry after the cutoff, then we ignore that row but continue
+		// examining the rest, as it is unfortunately possible for AUTO_INCREMENT
+		// columns to be assigned in non-monotonically-increasing order. These rows
+		// are only considered to be out-of-order if an in-order row is subsequently
+		// found later in the same batch.
 		if v.Expires.After(cutoff) {
-			break
+			oooRows++
```

Comment:
In the case of the last element of `data` being an out of order row, won't this cause the stat not to be incremented because it continues before that happens?

Reply:
This comment was meant to cover lines 236-244

Reply:
Yep, that's correct and intentional. The idea is that if the last element of data is an "out of order" row, then that's more like "in order" -- at some point we _expect_ the dates to move forward to dates that we don't care about (yet), and so this only counts rows as out-of-order if we see a date that we do care about _afterwards_.

Yes, this is a possibly-broken heuristic, if we just happen to hit the case where the one out-of-order row comes right at the end of a batch. But honestly I think the fix here is "close this PR, delete the janitor instead".